scrypto/src/resource/proof_rule.rs

```diff
@@ -1,94 +1,139 @@
 use crate::resource::*;
 use crate::rust::vec;
 use crate::rust::vec::Vec;
+use sbor::path::SborRelPath;
 use sbor::*;
 use scrypto::math::Decimal;
 
-/// Authorization Rule
 #[derive(Debug, Clone, PartialEq, Eq, Hash, Describe, TypeId, Encode, Decode)]
-pub enum ProofRule {
+pub struct SborPath(Vec<usize>);
+
+impl SborPath {
+    pub fn rel_path(&self) -> SborRelPath {
+        SborRelPath::new(&self.0)
+    }
+}
+
+impl From<&str> for SborPath {
+    fn from(str: &str) -> Self {
+        let path: Vec<usize> = str
+            .split('/')
+            .map(|s| s.parse::<usize>().unwrap())
+            .collect();
+        SborPath(path)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Describe, TypeId, Encode, Decode)]
+pub enum ProofRuleResource {
     NonFungible(NonFungibleAddress),
-    AnyOfResource(ResourceDefId),
-    SomeOfResource(Decimal, ResourceDefId),
-    AllOf(Vec<ProofRule>),
-    OneOf(Vec<ProofRule>),
-    CountOf { count: u8, rules: Vec<ProofRule> },
-}
-
-impl ProofRule {
-    pub fn or(self, other: ProofRule) -> Self {
-        match self {
-            ProofRule::NonFungible(_) => ProofRule::OneOf(vec![self, other]),
-            ProofRule::AnyOfResource(_) => ProofRule::OneOf(vec![self, other]),
-            ProofRule::SomeOfResource(_, _) => ProofRule::OneOf(vec![self, other]),
-            ProofRule::AllOf(_) => ProofRule::OneOf(vec![self, other]),
-            ProofRule::OneOf(mut rules) => {
-                rules.push(other);
-                ProofRule::OneOf(rules)
-            }
-            ProofRule::CountOf { count: _, rules: _ } => ProofRule::OneOf(vec![self, other]),
-        }
+    Resource(ResourceDefId),
+    FromComponent(SborPath),
+}
+
+impl From<NonFungibleAddress> for ProofRuleResource {
+    fn from(non_fungible_address: NonFungibleAddress) -> Self {
+        ProofRuleResource::NonFungible(non_fungible_address)
+    }
+}
+
+impl From<ResourceDefId> for ProofRuleResource {
+    fn from(resource_def_id: ResourceDefId) -> Self {
+        ProofRuleResource::Resource(resource_def_id)
     }
 }
 
+impl From<SborPath> for ProofRuleResource {
+    fn from(path: SborPath) -> Self {
+        ProofRuleResource::FromComponent(path)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Describe, TypeId, Encode, Decode)]
+pub enum ProofRuleResourceList {
+    StaticList(Vec<ProofRuleResource>),
+    FromComponent(SborPath),
+}
+
+impl From<SborPath> for ProofRuleResourceList {
+    fn from(path: SborPath) -> Self {
+        ProofRuleResourceList::FromComponent(path)
+    }
+}
+
+/// Authorization Rule
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Describe, TypeId, Encode, Decode)]
+pub enum ProofRule {
+    This(ProofRuleResource),
+    AmountOf(Decimal, ProofRuleResource),
+    CountOf(u8, ProofRuleResourceList),
+    AllOf(ProofRuleResourceList),
+    AnyOf(ProofRuleResourceList),
+}
```

Comment:
Minor: is it possible to make the model names consistent with the macro names, unless you want to introduce another layer of abstraction?

Reply:
yeah, gonna keep this like this for now as still unclear number of layers of abstraction needed.